### ISSUE TITLE
Rule request: has-own-property

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#3822]
#### TypeScript code being linted

```typescript
const container = { a: "a", b: "b", hasOwnProperty: () => false };

for (const i in container) {
    if (container.hasOwnProperty(i)) {
        console.log(container[i]);
    }
}
```

#### Expected behavior

`container.hasOwnProperty` isn't guaranteed to work as expected, since the object might override it. You're better off using `{}.hasOwnProperty.call(container, i)` (or, [as TypeScript prefers](https://github.com/Microsoft/TypeScript/issues/23208), `Object.prototype.hasOwnProperty.call(container, i)`.

Docs of this rule should probably also recommend the user consider using a `Map`.

**Comment by grimly** [comment]
You can use the [`ban`](https://palantir.github.io/tslint/rules/ban/) rule to that effect then make use of something like `Object.getOwnPropertyDescriptor(container, i) !== undefined` (you can even omit the `!== undefined` part)

**Comment by JoshuaKGoldberg** [comment]
Ooh, that should work for most scenarios. 

One corner case: if someone coincidentally has a member function named `hasOwnProperty` and wants to use it. That seems like a dangerous idea. Adding an example to the `ban` rule that explains both the reasons to ban `hasOwnProperty` and why you shouldn't name your methods that should be good.
